kratos/containers/csr_matrix.h

```diff
@@ -0,0 +1,587 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Riccardo Rossi
+//
+
+#if !defined(KRATOS_CSR_MATRIX_H_INCLUDED )
+#define  KRATOS_CSR_MATRIX_H_INCLUDED
+
+
+// System includes
+#include <iostream>
+#include "containers/sparse_contiguous_row_graph.h"
+#include "containers/system_vector.h"
+#include "utilities/parallel_utilities.h"
+#include "utilities/atomic_utilities.h"
+#include "includes/key_hash.h"
+
+// Project includes
+#include "includes/define.h"
+
+
+namespace Kratos
+{
+///@addtogroup ApplicationNameApplication
+///@{
+
+///@name Kratos Globals
+///@{
+
+///@}
+///@name Type Definitions
+///@{
+
+///@}
+///@name  Enum's
+///@{
+
+///@}
+///@name  Functions
+///@{
+
+///@}
+///@name Kratos Classes
+///@{
+
+/// This function implements "serial" CSR matrix, including capabilities for FEM assembly
+template< class TDataType=double, class TIndexType=std::size_t>
+class CsrMatrix
+{
+public:
+    ///@name Type Definitions
+    ///@{
+    typedef TIndexType IndexType;
+    typedef std::unordered_map<std::pair<IndexType, IndexType>,
+                          double,
+                          PairHasher<IndexType, IndexType>,
+                          PairComparor<IndexType, IndexType>
+                          > MatrixMapType;
+
+    /// Pointer definition of CsrMatrix
+    KRATOS_CLASS_POINTER_DEFINITION(CsrMatrix);
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+    CsrMatrix()
+    {
+    }
+
+    /// Default constructor.
+    template<class TGraphType>
+    CsrMatrix(const TGraphType& rSparseGraph)
+    {
+        rSparseGraph.ExportCSRArrays(mRowIndices,mColIndices);
+        mNrows = size1();
+
+        ComputeColSize();
+
+        //initialize mValuesVector to zero
+        mValuesVector.resize(mColIndices.size(),false);
+        SetValue(0.0);
+    }
+
+    /// Destructor.
+    virtual ~CsrMatrix(){}
+
+    /// Assignment operator. TODO: decide if we do want to allow it
+    CsrMatrix& operator=(CsrMatrix const& rOther)=delete;
```

Comment:
Just asking, which is the reasoning behind deleting the copy constructor? Don't you think it could be useful to keep it?

Reply:
the copy constructor makes too easy to make a copy, which is a quite expensive operation. 
There is no difficulty in implementing it ... simply i would rather not...

Reply:
So the reasoning is mainly to avoid accidental copies. My point is that it becomes tedious to create a copy in case it is needed.

Reply:
in @KratosMultiphysics/technical-committee  we decided to implement it as private